xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go

```diff
@@ -0,0 +1,558 @@
+
+	// Start backend servers which provide an implementation of the TestService.
+	servers, cleanup2 := startTestServiceBackends(t, 4)
+	defer cleanup2()
+	addrs, ports := backendAddressesAndPorts(t, servers)
+
+	const cacheTimeout = 100 * time.Microsecond
+	oldCacheTimeout := balancergroup.DefaultSubBalancerCloseTimeout
+	balancergroup.DefaultSubBalancerCloseTimeout = cacheTimeout
+	defer func() { balancergroup.DefaultSubBalancerCloseTimeout = oldCacheTimeout }()
+
+	// Create xDS resources for consumption by the test. The first update is an
+	// empty update. This should put the channel in TRANSIENT_FAILURE.
+	resources := clientResources(nodeID, edsServiceName, nil)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := managementServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Create an xDS client for use by the clusterresolver LB policy.
+	client, err := xdsclient.NewWithBootstrapContentsForTesting(bootstrapContents)
+	if err != nil {
+		t.Fatalf("Failed to create xDS client: %v", err)
+	}
+	defer client.Close()
+
+	// Create a manual resolver and push service config specifying the use of
+	// the cluster_resolver LB policy with a single discovery mechanism.
+	r := manual.NewBuilderWithScheme("whatever")
+	jsonSC := fmt.Sprintf(`{
+			"loadBalancingConfig":[{
+				"cluster_resolver_experimental":{
+					"discoveryMechanisms": [{
+						"cluster": "%s",
+						"type": "EDS",
+						"edsServiceName": "%s"
+					}],
+					"xdsLbPolicy":[{"round_robin":{}}]
+				}
+			}]
+		}`, clusterName, edsServiceName)
+	scpr := internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(jsonSC)
+	r.InitialState(xdsclient.SetClient(resolver.State{ServiceConfig: scpr}, client))
+
+	// Create a ClientConn and ensure TRANSIENT_FAILURE.
+	cc, err := grpc.Dial(r.Scheme()+":///test.service", grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithResolvers(r))
+	if err != nil {
+		t.Fatalf("failed to dial local test server: %v", err)
+	}
+	defer cc.Close()
+	testClient := testpb.NewTestServiceClient(cc)
+	if err := waitForTransientFailure(ctx, t, testClient); err != nil {
+		t.Fatal(err)
+	}
+
+	// Add a locality with one backend and ensure RPCs are successful.
+	resources = clientResources(nodeID, edsServiceName, []localityInfo{{name: localityName1, weight: 1, ports: ports[:1]}})
+	if err := managementServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+	if err := rrutil.CheckRoundRobinRPCs(ctx, testClient, addrs[:1]); err != nil {
+		t.Fatal(err)
+	}
+
+	// Push another empty update and ensure TRANSIENT_FAILURE.
+	resources = clientResources(nodeID, edsServiceName, nil)
+	if err := managementServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+	if err := waitForTransientFailure(ctx, t, testClient); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func waitForTransientFailure(ctx context.Context, t *testing.T, client testgrpc.TestServiceClient) error {
```

Comment:
Nit: are we really waiting for transient failure here specifically? I feel like the ClientConn can go TRANSIENT_FAILURE in an orthogonal situation than this specific error defined in the priority package. Can you please change name/add top level comment for function.